### ISSUE TITLE
BBB-PLAYBACK page not loading...

### ISSUE THREAD

**Issue by Kernelogy** [bigbluebutton/bbb-playback#117]
Hi BBB Team,

We are providing link for our users to view the recorded BBB presentations which are loaded through a custom domain, and the recording files are hosted in AWS-S3

An example url: 
https://my-domain/playback/presentation/2.3/meeting-id 

The BBB-Playback module is woking fine and presentations can be viewed without any problem when the url is loaded for the first time. 

But on loading the same url for the second time the BBB-playback module keeps on loading indefinitely and the presentation cannot be viewed.

But when we force reload(Ctrl+F5) the page, presentation is loading nicely without any issues. We are not able to locate the problem exactly but we can say so, the problem somehow related to session cache. 

Console Error info from the first load, this doesnot prevents the page from loading

![load-1](https://user-images.githubusercontent.com/29398353/125585267-41674b38-a805-462f-9f10-f4810fed82f2.png)

Console Error info from the second load and the page keeps on loading indefinitely. Unless we  force reload the page

![load-2](https://user-images.githubusercontent.com/29398353/125585304-ff869dcb-d79b-49a5-8f88-a06e58b99bb6.png)

Thanks & Regards

**Comment by pedrobmarin** [comment]
Hi,

Maybe check on your custom setup if you have everything right. Your logs seems very clear on the `Access-Control-Allow-Origin` presence (at least).

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Origin

**Comment by Kernelogy** [comment]
@pedrobmarin if that is the case the page should be loading for the first time itself. But on the loading other meetings which is not of the same meeting-id are also loading for the first time one after the other without any issues. Only when the page is loaded for the second or more than one time the issue is there. This seems bit tricky.

Also this is the case in all the browsers indefinite of the devices whether it desktop/laptop/mobile.

**Comment by pedrobmarin** [comment]
Well, from what I see there's nothing to be done at the playback code to fix that one for you. I suggest you at least try to set the `Access-Control-Allow-Origin`.

**Comment by Kernelogy** [comment]
Dear @pedrobmarin Please take a look at the below snapshot 

![load-3](https://user-images.githubusercontent.com/29398353/125942285-be0e6f79-eba4-4cb0-812e-20a8b25c8747.png)

All the presentation files are loading fine except the notes.html and webcams.mp4. As we didnt used notes in the meeting there is no notes.html file is present and we are using the webm format for the camera recordings so webcams.mp4 file is also not present along with the presentation data. 

There were no such files notes.html & webcams.mp4 were present in the meeting recordings and when the call is placed for these files which is not there it is returning 403. 

This has noting to do with the issue. The presentations are loading and We are able to view the presentation when the url is loaded for the first time. But on refreshing the page or open the same url in another browser tab the page is not loading. 

But when loading a diffrent url, with different meeting-id the same thing is happening. Loading for the first time and afterwards it is not loading. When we force reload(Ctrl+F5) the url is loading fine.

**Comment by Kernelogy** [comment]
After a long search and making various checkings we recompiled the bbb-playback code form the source and deployed bbb-playback it fixed the problem. 

Still we dont know what the actual problem is

Now it is showing the error code 404 for the missing files 
takers.json, webcam.mp4 and notes.html files are the missing files.

**Comment by Kernelogy** [comment]
**Found the issue and the solution at last**

The browser is making OPTIONS preflight request to the S3 for the resources. When the resources are loaded for the first time there is no issue arising out of the preflight requests. 

But once the resources are loaded they stays in the browser cache. So after that whenever the preflight request is made while the resources are in cache the preflight requests are failing. This is the case with Chrome/Chromium browsers. 

Even though the resources are in cache the preflight requests made by Firefox completes successfully. This _**cache vs cors preflight**_ nightmare is not affecting the Firefox that is for sure. 

A solution we applied for this is

We added  `Cache-Control: max-age=0, no-cache, no-store, must-revalidate` metadata for all the S3 Objects so that the resources are not cached in the browsers. 

And this is working.
